### PR TITLE
Implement partial specialization behind a flag

### PR DESCRIPTION
**Pull request by swiftix**
Partial specialization is disabled by default. Use -sil-partial-specialization to enable it.

Use -sil-partial-specialization-with-generic-substitutions to enable the partial specialization even in cases of substitutions containing generic replacement types.

### CODE REVIEW

**Comment by swiftix** [comment]
@swift-ci please smoke test

**Comment by swiftix** [comment]
@swift-ci please smoke test

**Comment by swiftix** [comment]
@swift-ci please smoke test

**Comment by swiftix** [comment]
@slavapestov @DougGregor Do you mind taking a look? It contains the implementation, which is now cleaner and more understandable. But it doesn't quite work yet due to the issues with the GenericSignatureBuilder.

The whole functionality is behind the flag and is disabled by default. So, it should be safe to merge. And it would solve the issue of going out of sync with any upcoming changes related to GenericSignatureBuilder and generics. 

So, OK to merge?

**Comment by slavapestov** [comment]
Please don't merge until I review.

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 1]
```diff
@@ -93,46 +124,183 @@ ReabstractionInfo::ReabstractionInfo(ApplySite Apply, SILFunction *OrigF,
...
 
-  SILModule &M = OrigF->getModule();
-  SubstitutedType = OrigF->getLoweredFunctionType()->substGenericArgs(
-    M, InterfaceSubs);
+  // Check if we have substitutions which replace generic type parameters with
+  // concrete types or unbound generic types.
+  bool HasConcreteGenericParams = false;
+  bool HasNonArchetypeGenericParams = false;
+  HasUnboundGenericParams = false;
+  for (auto DT : CalleeGenericSig->getAllDependentTypes()) {
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: You can just call getGenericParams() here

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 2]
```diff
@@ -93,46 +124,183 @@ ReabstractionInfo::ReabstractionInfo(ApplySite Apply, SILFunction *OrigF,
...
+    // Check only the substitutions for the generic parameters.
+    // Ignore any dependent types, etc.
+    if (!DT->is<GenericTypeParamType>())
+      continue;
+    auto Replacement = InterfaceSubs
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: This is better expressed with Type(DT).subst(InterfaceSubs)

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 3]
```diff
@@ -93,46 +124,183 @@ ReabstractionInfo::ReabstractionInfo(ApplySite Apply, SILFunction *OrigF,
...
+                           .lookupSubstitution(cast<GenericTypeParamType>(
+                               DT->getCanonicalType()))
+                           ->getCanonicalType();
+    if (!isa<ArchetypeType>(Replacement) &&
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: You shouldn't have both archetypes and interface types in the map.

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 4]
```diff
@@ -93,46 +124,183 @@ ReabstractionInfo::ReabstractionInfo(ApplySite Apply, SILFunction *OrigF,
...
+    if (Replacement->hasArchetype()) {
+      HasUnboundGenericParams = true;
+      // Check if the replacement is an archetype which is more specific
+      // than the corresponding archetype in the original generic signature.
+      // If this is the case, then specialization makes sense, because
+      // it would produce something more specific.
+      if (CalleeGenericEnv) {
+        if (auto Archetype = dyn_cast<ArchetypeType>(Replacement)) {
+          auto OrigArchetype = dyn_cast<ArchetypeType>(
+              CalleeGenericEnv->mapTypeIntoContext(DT)->getCanonicalType());
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: Use foo->getAs<ArchetypeType>() instead of dyn_cast<ArchetypeType>(foo->getCanonicalType())

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 5]
```diff
@@ -280,6 +456,25 @@ createSpecializedType(CanSILFunctionType SubstFTy, SILModule &M) const {
...
 }
 
+static std::pair<GenericEnvironment *, GenericSignature *>
+getGenericEnvironmentAndSignature(GenericSignatureBuilder &Builder,
+                                  SILModule &M) {
+  auto *GenericSig =
+      Builder.getGenericSignature()->getCanonicalSignature().getPointer();
+  auto *GenericEnv = GenericSig->createGenericEnvironment(*M.getSwiftModule());
+
+  // FIXME: This is a workaround for the signature minimization bug.
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: Since the flag disables the entire optimization, please remove the workaround. We will fix it before enabling

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 6]
```diff
@@ -399,6 +587,683 @@ ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
...
+    verifySubstitutionList(ClonerParamSubs);
+  }
+
+  {
+    SmallVector<Substitution, 4> List;
+
+    SpecializedGenericSig->getSubstitutions(
+      [&](SubstitutableType *type) -> Type {
+        //return OrigGenericEnv->mapTypeIntoContext(type);
+        return InterfaceSubs.lookupSubstitution(cast<SubstitutableType>(type->getCanonicalType()));
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: Again use subst() instead of lookupSubstitution(), then you don't need to cast or canonicalize and it's also more correct

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 7]
```diff
@@ -399,6 +587,683 @@ ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
...
+
+    if (isa<DependentMemberType>(Replacement->getCanonicalType()))
+      return false;
+  }
+
+  if (OptimizeGenericSubstitutions) {
+    // Is it an unconstrained generic parameter?
+    if (Conformances.empty()) {
+      if (Replacement->is<ArchetypeType>() ||
+          Replacement->is<DependentMemberType>()) {
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: Again, why the weird archetype or member type check?

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 8]
```diff
@@ -399,6 +587,683 @@ ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
...
+}
+
+
+/// Collect all used archetypes from all the substitutions.
+/// Take into account only those archetypes that occur in the
+/// substitutions of generic parameters which will be partially
+/// specialized. Ignore all others.
+static void
+callectUsedArchetypes(ArrayRef<Substitution> ParamSubs,
+                      llvm::SmallSetVector<CanType, 8> &UsedCallerArchetypes) {
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: Make this a vector of ArchetypeType* instead

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 9]
```diff
@@ -399,6 +587,683 @@ ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
...
+
+  for (auto Sub : ParamSubs) {
+    auto Replacement = Sub.getReplacement()->getCanonicalType();
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: More redundant getCanonicalType() calls

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 10]
```diff
@@ -399,6 +587,683 @@ ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
...
+
+    // If the substitution will not be performed in the specialized
+    // function, there is no need to check for any archetypes inside
+    // the replacement.
+    if (!shouldBePartiallySpecialized(Replacement, Sub.getConformances()))
+      continue;
+
+    // Add used generic parameters/archetypes.
+    Replacement.visit([&](Type Ty) {
+        if (auto Archetype = dyn_cast<ArchetypeType>(Ty->getCanonicalType())) {
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: Again use getAs instead of dyn_cast + canonicalize

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 11]
```diff
@@ -399,6 +587,683 @@ ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
...
+  auto BaseTy = Ty;
+  while (auto DepTy = dyn_cast<DependentMemberType>(BaseTy)) {
+    BaseTy = DepTy->getBase()->getCanonicalType();
+  };
+  assert(isa<GenericTypeParamType>(BaseTy));
+  return cast<GenericTypeParamType>(BaseTy);
+}
+
+/// Replace dependent types with their archetypes or concrete types.
+static Type substConcreteTypesForDependentTypes(
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: This is another workaround that I thought we already agreed to remove

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 12]
```diff
@@ -399,6 +587,683 @@ ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
...
+
+  unsigned Depth = 0;
+  unsigned Index = 0;
+
+  // Add generic parameters that will come from the callee.
+  for (auto GP : CalleeGenericSig->getGenericParams()) {
+    auto CanTy = GP->getCanonicalType();
+    Type Replacement = CalleeGenericSig->getConcreteType(CanTy, *SM);
+    if (!Replacement)
+      Replacement = CalleeInterfaceToCallerArchetypeMap.lookupSubstitution(
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: Again use subst()

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 13]
```diff
@@ -399,6 +587,683 @@ ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
...
+      Builder.addRequirement(Req, Source);
+    }
+  }
+
+  // Generate a new generic type parameter for each used archetype from
+  // the caller.
+  for (auto CallerArchetype : UsedCallerArchetypes) {
+    auto CallerGenericParam =
+        CallerGenericEnv->mapTypeOutOfContext(CallerArchetype)
+            ->getCanonicalType();
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: Why canonicalize here?

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 14]
```diff
@@ -399,6 +587,683 @@ ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
...
+      CalleeGenericSig->getSubstitutionMap(ParamSubs);
+
+  DEBUG(llvm::dbgs() << "\n\nTrying partial specialization for: "
+                     << Callee->getName() << "\n";
+        llvm::dbgs() << "Callee generic signature is:\n";
+        CalleeGenericSig->dump());
+
+  // Maps caller's generic parameters to generic parameters of the specialized
+  // function.
+  llvm::DenseMap<CanType, Type>
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: Why are these needed and why aren't they SubstitutionMaps

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 15]
```diff
@@ -399,6 +587,683 @@ ReabstractionInfo::ReabstractionInfo(SILFunction *OrigF,
...
+
+
+  // Next, add each of the requirements (mapped from the requirement's
+  // interface types into the specialized interface type parameters).
+  // TODO: Do we need to add requirements of the caller's archetypes, which
+  // stem from the caller's generic signature? If so, which ones? All of them?
+  // Just some of them? Most likely we need to add only those which are not
+  // present in the callee's signature.
+
+  auto MapCalleeInterfaceTypeToSpecializedInterfaceType = [&](Type ty) -> Type {
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: This is really ugly and not using the right abstractions

**Review by slavapestov — CHANGES_REQUESTED** [review]

File: lib/SILOptimizer/Utils/Generics.cpp (hunk 5 above)
Comment (swiftix) [inline lib/SILOptimizer/Utils/Generics.cpp]: I want to keep it so that you or Doug can test thing when you change something e.g. in the GenericSignatureBuilder

**Comment by swiftix** [comment]
@slavapestov I addressed most of your comments. Please have a look.

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 16]
```diff
@@ -772,9 +748,9 @@ shouldBePartiallySpecialized(Type Replacement,
...
 /// specialized. Ignore all others.
-static void
-callectUsedArchetypes(ArrayRef<Substitution> ParamSubs,
-                      llvm::SmallSetVector<CanType, 8> &UsedCallerArchetypes) {
+static void callectUsedArchetypes(
+    ArrayRef<Substitution> ParamSubs,
+    llvm::SmallSetVector<ArchetypeType *, 2> &UsedCallerArchetypes) {
 
   for (auto Sub : ParamSubs) {
     auto Replacement = Sub.getReplacement()->getCanonicalType();
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: Don't need to canonicalize here

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 17]
```diff
@@ -1009,8 +954,7 @@ void ReabstractionInfo::specializeConcreteAndGenericSubstitutions(
     auto CanTy = GP->getCanonicalType();
     Type Replacement = CalleeGenericSig->getConcreteType(CanTy, *SM);
     if (!Replacement)
-      Replacement = CalleeInterfaceToCallerArchetypeMap.lookupSubstitution(
-          cast<SubstitutableType>(CanTy));
+      Replacement = CanTy.subst(CalleeInterfaceToCallerArchetypeMap);
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: Don't need to canonicalize here either

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 18]
```diff
@@ -1175,11 +1099,20 @@ void ReabstractionInfo::specializeConcreteAndGenericSubstitutions(
...
+            LookUpConformanceInSignature(*SpecializedGenericSig));
+
     SpecializedGenericSig->getSubstitutions(
         [&](SubstitutableType *type) -> Type {
           auto Replacement =
-              SpecializedInterfaceToCallerArchetypeMapping.lookup(
-                  type->getCanonicalType());
+              Type(type).subst(SpecializedInterfaceToCallerArchetypeMap);
           if (Replacement)
             return Replacement;
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: Is it possible for this to fail still? If not, this whole part can just become SpecializedGenericSig->getSubstitutions(SpecializedInterfaceToCallerArchetypeMap, List);

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 19]
```diff
@@ -1219,7 +1152,7 @@ void ReabstractionInfo::specializeConcreteAndGenericSubstitutions(
         M,
         [&](SubstitutableType *type) -> Type {
           auto SpecializedInterfaceTy =
-              CalleeInterfaceToSpecializedInterfaceType(type);
+              MapCalleeInterfaceTypeToSpecializedInterfaceType(type);
 
           return SpecializedGenericEnv->mapTypeOutOfContext(
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: The map out/into context should not be necessary. In fact, this can just be SpecializedSubstFnTy = CalleeFnTy->substGenericArgs(M, SpecializedInterfaceToCallerArchetypeMap)

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 20]
```diff
@@ -960,7 +906,6 @@ void ReabstractionInfo::specializeConcreteAndGenericSubstitutions(
   // Callee is the generic function being called by the apply instruction.
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: Can you break up this function a bit into pieces that compute different things?

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 21]
```diff
@@ -1135,35 +1081,13 @@ void ReabstractionInfo::specializeConcreteAndGenericSubstitutions(
...
-          return Ty;
-        },
-        LookUpConformanceInSignature(*SpecializedGenericSig));
-    return SpecializedInterfaceTy;
-  };
-
   {
     SmallVector<Substitution, 4> List;
 
     CalleeGenericSig->getSubstitutions(
```
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: This should just be CalleeGenericSig->GetSubstitutions(CallerInterfaceToSpecializedInterfaceMap, List)

Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: Oh, never mind -- there's a mapTypeIntoContext there.

File: lib/SILOptimizer/Utils/Generics.cpp (hunk 19 above)
Comment (swiftix) [inline lib/SILOptimizer/Utils/Generics.cpp]: I'm not sure about this. We need to gat an interface type at the end, right? If we use SpecializedInterfaceToCallerArchetypeMap, we are going to get a contextual type, or?

File: lib/SILOptimizer/Utils/Generics.cpp (hunk 17 above)
Comment (swiftix) [inline lib/SILOptimizer/Utils/Generics.cpp]: What do you mean here? `CanTy`? It is done because later it is used for `getConformances` which require CanType.

File: lib/SILOptimizer/Utils/Generics.cpp (hunk 18 above)
Comment (swiftix) [inline lib/SILOptimizer/Utils/Generics.cpp]: Hmm. I guess you are right. The only way how it could fail is when a substitution is concrete, but in this case it is known already and the generic signature would contain `T == ConcreteType` for it and thus getSubstitutions will not ask for it, right? Because it only collects substitutions which are really supposed to be provided by an apply.

File: lib/SILOptimizer/Utils/Generics.cpp (hunk 20 above)
Comment (swiftix) [inline lib/SILOptimizer/Utils/Generics.cpp]: I tried to create separate functions. But they need to share to much state. So, for now I added big comments delimiting different logical sections.

**Comment by swiftix** [comment]
@slavapestov As for your hack. It works for me. I can remove the workaround on my side if you are going to merge your hack.

File: lib/SILOptimizer/Utils/Generics.cpp (hunk 5 above)
Comment (slavapestov) [inline lib/SILOptimizer/Utils/Generics.cpp]: We plan on adding an assert soon to make sure canonicalized signatures are minimal, so we won't need this check

**Comment by slavapestov** [comment]
I'll merge the hack and the other PR tonight. Then we can merge this I think.

File: lib/SILOptimizer/Utils/Generics.cpp (hunk 5 above)
Comment (swiftix) [inline lib/SILOptimizer/Utils/Generics.cpp]: That's OK. I'm fine removing it once you introduce this check in the GenericSignatureBuilder (i.e. you check that it is minimal and idempotent)

**Comment by slavapestov** [comment]
Time to merge this?

**Comment by swiftix** [comment]
@slavapestov  Simplified version of it (without generic substitutions) is merged already, see #7888

**Comment by swiftix** [comment]
This one is obsolete now.

**Comment by swiftix** [comment]
I'm going to close it. I'll submit a new PR implementing a more general approach, once it is ready and GenericSignatureBuilder issues are fixed.
